mailbox/plugin/quota-search/src/test/java/org/apache/james/quota/search/QuotaClauseTest.java

```diff
@@ -0,0 +1,121 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.List;
+
+import org.apache.james.core.Domain;
+import org.junit.jupiter.api.Test;
+
+import nl.jqno.equalsverifier.EqualsVerifier;
+
+public class QuotaClauseTest implements QuotaClause {
+    
+    @Test
+    public void lessThanShouldMatchBeanContract() {
+        EqualsVerifier.forClass(QuotaClause.LessThan.class)
+            .allFieldsShouldBeUsed()
+            .verify();
+    }
+    
+    @Test
+    public void moreThanShouldMatchBeanContract() {
+        EqualsVerifier.forClass(QuotaClause.MoreThan.class)
+            .allFieldsShouldBeUsed()
+            .verify();
+    }
+    
+    @Test
+    public void hasDomainShouldMatchBeanContract() {
+        EqualsVerifier.forClass(QuotaClause.HasDomain.class)
+            .allFieldsShouldBeUsed()
+            .verify();
+    }
+    
+    @Test
+    public void andShouldMatchBeanContract() {
+        EqualsVerifier.forClass(QuotaClause.And.class)
+            .allFieldsShouldBeUsed()
+            .verify();
+    }
+
+    @Test
+    public void lessThanClauseShouldThrowWhenQuotaThresholdIsNull() {
+        assertThatThrownBy(() -> QuotaClause.lessThan(null))
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void moreThanClauseShouldThrowWhenQuotaThresholdIsNull() {
+        assertThatThrownBy(() -> QuotaClause.moreThan(null))
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void hasDomainClauseShouldThrowWhenDomainIsNull() {
+        assertThatThrownBy(() -> QuotaClause.hasDomain(null))
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void andClauseShouldThrowWhenDomainIsNull() {
+        List<QuotaClause> clauses = null;
+        assertThatThrownBy(() -> QuotaClause.and(clauses))
+            .isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    public void lessThanClauseShouldInstanciate() {
+        QuotaClause.LessThan lessThanQuery = QuotaClause.lessThan(_50);
+
+        assertThat(lessThanQuery.getQuotaBoundary()).isEqualTo(_50);
+    }
+
+    @Test
+    public void moreThanClauseShouldInstanciate() {
+        QuotaClause.MoreThan moreThanQuery = QuotaClause.moreThan(_50);
+
+        assertThat(moreThanQuery.getQuotaBoundary()).isEqualTo(_50);
+    }
+
+    @Test
+    public void hasDomainClauseShouldInstanciate() {
+        Domain domain = Domain.of("domain.org");
+
+        QuotaClause.HasDomain hasDomainQuery = QuotaClause.hasDomain(domain);
+
+        assertThat(hasDomainQuery.getDomain()).isEqualTo(domain);
+    }
+
+    @Test
+    public void andClauseShouldInstanciate() {
+        QuotaClause.MoreThan first = QuotaClause.moreThan(_50);
+        QuotaClause.MoreThan second = QuotaClause.moreThan(_75);
+
+        QuotaClause.And andQuery = QuotaClause.and(first, second);
+
+        assertThat(andQuery.getClauses())
+            .containsExactly(first, second);
+    }
```

Comment:
what about nested and clauses? are they supported? (it seems so)

Reply:
They are supported by the DSL but not the DSL builder, **as allowing it brings complexity and is not a requirement**.

I'm not sure it makes sens to add such tests then (but I can)

Reply:
you can just add a test at this level to show that's it ok, or add a constraint so that no clause of an and clause is an and clause (as it not supported in other levels)

Reply:
yes it seems the better choice :+1: